afew/filters/DKIMValidityFilter.py

```diff
@@ -49,7 +50,23 @@ def __init__(self, database, ok_tag='dkim-ok', fail_tag='dkim-fail'):
         self.log = logging.getLogger('{}.{}'.format(
             self.__module__, self.__class__.__name__))
 
-    def handle_message(self, message):
+    async def run(self, query):
+        self.log.info(self.message)
+
+        if getattr(self, "query", None):
+            if query:
+                query = "(%s) AND (%s)" % (query, self.query)
+            else:
+                query = self.query
+
+        coroutines = set()
+
+        for message in self.database.get_messages(query):
+            coroutines.add(self.handle_message(message))
+
+        return await asyncio.gather(*coroutines, return_exceptions=True)
```

Comment:
Why not implement this in a base class, e.g. AsyncBaseFilter? It would be easier to migrate the other filters. The change required in a filter implementation is simple enough, would could easily migrate all the filters and avoid duplicate code (like part of this method vs BaseFilter) and hack, like the `iscoroutinefunction()` check.